### ISSUE TITLE
Wrap model variables with xarray

### ISSUE THREAD

**Issue by rabernat** [pyqg/pyqg#145]
It would be awesome to wrap the model variables with [xarray](http://xarray.pydata.org/en/stable/) data structures. This would make so many things easier for the user, including plotting and serialization to netCDF.

This would probably be a pretty major refactor.

**Comment by rabernat** [comment]
This would enable #9.

**Comment by rabernat** [comment]
Some more guidance about how this might work...

We should perhaps attach a method called `to_dataset()` to the pyqg `Model` object. The method will return an xarray dataset with all of the model prognostic variables and diagnostic variables in it.

### Dimensions & Coordinates

Every Xarray dataset needs to define its dimensions and coordinates. Pyqg has both spatial coordinates (`x`, `y`) and spectral ones (`k` and `l`), plus `level`. Variables are either spectral or spatial.

### Prognostic variables

The prognostic variables are those that are part of the kernel. They are the non-private variables enumerated here:

https://github.com/pyqg/pyqg/blob/master/pyqg/kernel.pyx#L32-L77

For each variable, we need to identify 
- its dims
- the data itself (easy)
- the attrs. We should add attributes like `units` and `standard_name` and / or `long_name` to the attrs, following [CF conventions](https://cfconventions.org/).

### Diagnostic variables

The diagnostic variables are messier.

https://github.com/pyqg/pyqg/blob/master/pyqg/model.py#L571-L597

We will need to figure out how to encode their dims and attrs. That could be done as a second pass, after first doing the prognostic vars.

**Comment by rabernat** [comment]
Creating a dataarray:

http://xarray.pydata.org/en/stable/data-structures.html#creating-a-dataarray

**Comment by rabernat** [comment]
Some pseudo code I wrote today

```python
spatial_dims = ('lev', 'y', 'x')
spectral_dims = ('lev', 'l', 'k')
dim_database = {
    'q': spatial_dims,
    'u': spatial_dims,
    'qh': spectral_dims,
    'ph': spectral_dims
    
}

variables = {}
for vname in dim_database:
    data = getattr(model, vname)
    variables[vname] = (dims, data)
    
ds = xr.Dataset(variables)
```

**Comment by hscannell** [comment]
This is awesome @paigem! Since we're working asynchronously, I can bite off a bit more work on this today.

**Comment by hscannell** [comment]
I slogged away on this for a bit today, building on top of the work @paigem has already done. I'm working from a forked branch [here](https://github.com/hscannell/pyqg/blob/wrap-xarray/examples/wrap-xarray.ipynb). 

**To Do List:**

- [ ] put the correct units in the attributes
- [ ] use the proper CF standard names
- [ ] are we missing more variables from other `pyqg` models? I've mainly looked at the `QGModel`.
- [ ] add model diagnostics
- [ ] ...



```<xarray.Dataset>
Dimensions:  (nk: 33, nl: 64, nx: 64, ny: 64, nz: 2)
Coordinates:
    z        (nz) int64 0 1
    x        (ny, nx) float64 7.812e+03 2.344e+04 ... 9.766e+05 9.922e+05
    y        (ny, nx) float64 7.812e+03 7.812e+03 ... 9.922e+05 9.922e+05
    l        (nl, nk) float64 0.0 0.0 0.0 ... -6.283e-06 -6.283e-06 -6.283e-06
    k        (nl, nk) float64 0.0 6.283e-06 1.257e-05 ... 0.0001948 0.0002011
    kk       (nk) float64 0.0 6.283e-06 1.257e-05 ... 0.0001948 0.0002011
    ll       (nl) float64 0.0 6.283e-06 1.257e-05 ... -1.257e-05 -6.283e-06
Dimensions without coordinates: nk, nl, nx, ny, nz
Data variables:
    q        (nz, ny, nx) float64 -1.279e-05 -6.909e-06 ... 2.017e-06 2.446e-06
    u        (nz, ny, nx) float64 0.03087 0.03018 ... -0.007114 -0.008991
    v        (nz, ny, nx) float64 -0.01866 -0.007716 ... 0.004994 0.01145
    ufull    (nz, ny, nx) float64 0.05587 0.05518 ... -0.007114 -0.008991
    vfull    (nz, ny, nx) float64 -0.01866 -0.007716 ... 0.004994 0.01145
    qh       (nz, nl, nk) complex128 (0.002008344377551905+0j) ... (-4.993187...
    uh       (nz, nl, nk) complex128 (-108.03894081433111-108.03894081433111j...
    vh       (nz, nl, nk) complex128 (-7.023273071743378e-18+7.02327307174337...
    ph       (nz, nl, nk) complex128 0j ... (1.9283370344075782e-16+3.6549036...
    a        (nz, nz, nl, nk) complex128 0j ... (-24223355.801269166+0j)
    Ubg      (nz) float64 0.025 0.0
    Qy       (nz) float64 1.039e-10 -7.222e-12
Attributes: (12/15)
    nx:         (64, {'standard_name': 'nx', 'description': 'number of real s...
    ny:         (64, {'standard_name': 'ny', 'description': 'number of real s...
    nl:         (64, {'standard_name': 'nl', 'description': 'number of spectr...
    rek:        (5.787e-07, {'standard_name': 'rek', 'description': 'linear d...
    t:          (311040000.0, {'standard_name': 'time', 'description': 'model...
    tc:         (43200, {'standard_name': 'tc', 'description': 'model timeste...
    ...         ...
    filterfac:  (23.6, {'standard_name': 'filterfac', 'description': 'amplitu...
    twrite:     (10000, {'standard_name': 'twrite', 'description': 'interval ...
    tmax:       (311040000.0, {'standard_name': 'time', 'description': 'total...
    tavestart:  (155520000.0, {'standard_name': 'tavestart', 'description': '...
    taveint:    (86400.0, {'standard_name': 'taveint', 'description': 'time i...
    ntd:        (1, {'standard_name': 'ntd', 'description': 'number of thread...```

**Comment by hscannell** [comment]
I'm also wondering if we need to put *everything* in the dataset, especially the attributes nx, ny, and nl.

**Comment by rabernat** [comment]
Some comments
- x, y should probably be the spatial dimensions (rather than nx, ny), and they should be made 1D
- same for k, l in the spectral dimension
- I see nx, ny, etc as constants, not coordinates. I see no harm in adding lots of attributes. The more metadata the better, no? Ideally the dataset should contain everything you need to know about how pyqg was run.

**Comment by hscannell** [comment]
x, y, k, l all look they are provided as a 2D meshgrid from pyqg. I had to reshape them in the dataset so now they are 1D. 
My thought was that nx, ny, nl, nk are essentially like len(x), len(y), etc. and seemed a bit excessive, but I've kept them in attributes for now. 

![Screen Shot 2021-03-11 at 10 02 07 AM](https://user-images.githubusercontent.com/22622585/110807306-e374a880-8250-11eb-944d-0d67d97efafc.png)

**Comment by rabernat** [comment]
This looks almost perfect. But whats the difference between `k` and `kk`? If they're both 1D, what's the point of having both?

My recollection was that kk and ll were 2D repeated versions of the basic 1D arrays that helped for some internal calculations. But xarray's broadcasting means we don't have to keep such arrays around.

Could you call `ds.info()` and paste the full output so we can see all the attributes for everything? It will be very long but that's okay.

**Comment by rabernat** [comment]
Oh, also, the following variable are not time dependent and shouldn't have a time dimension.
- `a`
- `Ubg`
- `Qy`

`a` is the inversion matrix. I'm not sure we want to expose it to users. I would leave it out of the dataset.

**Comment by hscannell** [comment]
Yes, you're totally right. I removed `kk` and `ll` from the variable list. I also removed `a` and fixed the dimensions of `Ubg` and `Qy`.

```
xarray.Dataset {
dimensions:
	k = 33 ;
	l = 64 ;
	time = 1 ;
	x = 64 ;
	y = 64 ;
	z = 2 ;

variables:
	float64 q(time, z, y, x) ;
		q:standard_name = q ;
		q:description = potential vorticity in real space ;
		q:units =  ;
	float64 u(time, z, y, x) ;
		u:standard_name = u ;
		u:description = zonal velocity anomaly ;
		u:units =  ;
	float64 v(time, z, y, x) ;
		v:standard_name = v ;
		v:description = meridional velocity anomaly ;
		v:units =  ;
	float64 ufull(time, z, y, x) ;
		ufull:standard_name = ufull ;
		ufull:description = zonal full velocities in real space ;
		ufull:units =  ;
	float64 vfull(time, z, y, x) ;
		vfull:standard_name = vfull ;
		vfull:description = meridional full velocities in real space ;
		vfull:units =  ;
	complex128 qh(time, z, l, k) ;
		qh:standard_name = qh ;
		qh:description = potential vorticity in spectral space ;
		qh:units =  ;
	complex128 uh(time, z, l, k) ;
		uh:standard_name = uh ;
		uh:description = zonal velocity anomaly in spectral space ;
		uh:units =  ;
	complex128 vh(time, z, l, k) ;
		vh:standard_name = vh ;
		vh:description = meridional velocity anomaly in spectral space ;
		vh:units =  ;
	complex128 ph(time, z, l, k) ;
		ph:standard_name = ph ;
		ph:description = streamfunction in spectral space ;
		ph:units =  ;
	float64 Ubg(z) ;
		Ubg:standard_name = Ubg ;
		Ubg:description = background zonal velocity ;
		Ubg:units =  ;
	float64 Qy(z) ;
		Qy:standard_name = Qy ;
		Qy:description = background potential vorticity gradient ;
		Qy:units =  ;
	float64 time(time) ;
		time:standard_name = time ;
		time:description = model time ;
		time:units = model time increment ;
	int64 z(z) ;
		z:standard_name = z ;
		z:description = vertical levels ;
		z:units = level ;
	float64 x(x) ;
		x:standard_name = x ;
		x:description = real space grid points in the x direction ;
		x:units = grid point ;
	float64 y(y) ;
		y:standard_name = y ;
		y:description = real space grid points in the y direction ;
		y:units = grid point ;
	float64 l(l) ;
		l:standard_name = l ;
		l:description = spectal space grid points in the l direction ;
		l:units = meridional wavenumbers ;
	float64 k(k) ;
		k:standard_name = k ;
		k:description = spectal space grid points in the k direction ;
		k:units = zonal wavenumbers ;

// global attributes:
	:nx = (64, {'standard_name': 'nx', 'description': 'number of real space grid points in x direction', 'units': ''}) ;
	:ny = (64, {'standard_name': 'ny', 'description': 'number of real space grid points in y direction (default: nx)', 'units': ''}) ;
	:nz = (2, {'standard_name': 'nz', 'description': 'number of vertical levels', 'units': ''}) ;
	:nl = (64, {'standard_name': 'nl', 'description': 'number of spectral space grid points in l direction', 'units': ''}) ;
	:rek = (5.787e-07, {'standard_name': 'rek', 'description': 'linear drag in lower layer', 'units': 'seconds^(-1)'}) ;
	:tc = (43200, {'standard_name': 'tc', 'description': 'model timestep', 'units': ''}) ;
	:dt = (7200.0, {'standard_name': 'st', 'description': 'numerical timestep', 'units': ''}) ;
	:L = (1000000.0, {'standard_name': 'L', 'description': 'domain length in x direction', 'units': ''}) ;
	:W = (1000000.0, {'standard_name': 'W', 'description': 'domain length in y direction', 'units': ''}) ;
	:filterfac = (23.6, {'standard_name': 'filterfac', 'description': 'amplitude of spectral spherical filter', 'units': ''}) ;
	:twrite = (10000, {'standard_name': 'twrite', 'description': 'interval for cfl writeout (units: number of timesteps)', 'units': ''}) ;
	:tmax = (311040000.0, {'standard_name': 'tmax', 'description': 'total time of integration', 'units': 'model time'}) ;
	:tavestart = (155520000.0, {'standard_name': 'tavestart', 'description': 'start time for averaging (units: model time)', 'units': ''}) ;
	:taveint = (86400.0, {'standard_name': 'taveint', 'description': 'time interval for accumulation of diagnostic averages (units: model time)', 'units': ''}) ;
	:ntd = (1, {'standard_name': 'ntd', 'description': 'number of threads used', 'units': ''}) ;
}
```

**Comment by rabernat** [comment]
Thanks for posting the full `info` dump. Based on this, I'm going to go back on what I said earlier (sorry! 😬) and suggest we move those scalar attributes _back to coordiantes_. The reason is that, unless they are actual dataset variables, they can't have the sort of metadata you've put in there. Right now you just have

`ds.nx = (64, {'standard_name': 'nx', 'description': 'number of real space grid points in x direction', 'units': ''})`

This is just a python tuple containing an integer and a dictionary. Xarray doesn't understand the relationships between those different things. In other words, attributes of attributes are not allowed

Just a few more things are needed to make this CF compliant
- Fill in the dataset-level attributes with the some of the [required CF metadata](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#description-of-file-contents), e.g. conventions, title, etc.
- change `description` to `long_name`
- remove `standard_name`. The names you are using are not [valid CF standard names](https://cfconventions.org/Data/cf-standard-names/77/build/cf-standard-name-table.html) and I don't think they exist for QG variables.

**Comment by hscannell** [comment]
That makes sense. I might need some help checking that these units are correct and filling in the appropriate dataset attributes. I will work on adding the diagnostic variables, but it might be helpful to discuss how this will eventually integrate in with `pyqg`. I assume we will want to create a function and call it within the model classes... 
My notebook lives [here](https://github.com/hscannell/pyqg/blob/wrap-xarray/examples/wrap-xarray.ipynb).

Here is an updated copy of ds.info() :

```
xarray.Dataset {
dimensions:
	k = 33 ;
	l = 64 ;
	time = 1 ;
	x = 64 ;
	y = 64 ;
	z = 2 ;

variables:
	float64 q(time, z, y, x) ;
		q:long_name = potential vorticity in real space ;
		q:units = m² s⁻¹ K kg⁻¹ ;
	float64 u(time, z, y, x) ;
		u:long_name = zonal velocity anomaly ;
		u:units = m s⁻¹ ;
	float64 v(time, z, y, x) ;
		v:long_name = meridional velocity anomaly ;
		v:units = m s⁻¹ ;
	float64 ufull(time, z, y, x) ;
		ufull:long_name = zonal full velocities in real space ;
		ufull:units = m s⁻¹ ;
	float64 vfull(time, z, y, x) ;
		vfull:long_name = meridional full velocities in real space ;
		vfull:units = m s⁻¹ ;
	complex128 qh(time, z, l, k) ;
		qh:long_name = potential vorticity in spectral space ;
		qh:units = m² s⁻¹ K kg⁻¹ ;
	complex128 uh(time, z, l, k) ;
		uh:long_name = zonal velocity anomaly in spectral space ;
		uh:units = m s⁻¹ ;
	complex128 vh(time, z, l, k) ;
		vh:long_name = meridional velocity anomaly in spectral space ;
		vh:units = m s⁻¹ ;
	complex128 ph(time, z, l, k) ;
		ph:long_name = streamfunction in spectral space ;
		ph:units = m² s⁻¹ ;
	float64 Ubg(z) ;
		Ubg:long_name = background zonal velocity ;
		Ubg:units = m s⁻¹ ;
	float64 Qy(z) ;
		Qy:long_name = background potential vorticity gradient ;
		Qy:units = m² s⁻¹ K kg⁻¹ ;
	float64 time(time) ;
		time:long_name = model time ;
		time:units = seconds ;
	int64 z(z) ;
		z:long_name = vertical levels ;
		z:units = none ;
	float64 x(x) ;
		x:long_name = real space grid points in the x direction ;
		x:units = grid point ;
	float64 y(y) ;
		y:long_name = real space grid points in the y direction ;
		y:units = grid point ;
	float64 l(l) ;
		l:long_name = spectal space grid points in the l direction ;
		l:units = meridional wavenumber ;
	float64 k(k) ;
		k:long_name = spectal space grid points in the k direction ;
		k:units = zonal wavenumber ;
	int64 nx() ;
		nx:long_name = number of real space grid points in x direction ;
		nx:units = none ;
	int64 ny() ;
		ny:long_name = number of real space grid points in y direction (default: nx) ;
		ny:units = none ;
	int64 nz() ;
		nz:long_name = number of vertical levels ;
		nz:units = none ;
	int64 nl() ;
		nl:long_name = number of spectral space grid points in l direction ;
		nl:units = grid point ;
	int64 nk() ;
		nk:long_name = number of spectral space grid points in k direction ;
		nk:units = grid point ;
	float64 rek() ;
		rek:long_name = linear drag in lower layer ;
		rek:units = seconds⁻¹ ;
	int64 tc() ;
		tc:long_name = model timestep ;
		tc:units = seconds ;
	float64 dt() ;
		dt:long_name = numerical timestep ;
		dt:units = seconds ;
	float64 L() ;
		L:long_name = domain length in x direction ;
		L:units = meters ;
	float64 W() ;
		W:long_name = domain length in y direction ;
		W:units = meters ;
	float64 filterfac() ;
		filterfac:long_name = amplitude of spectral spherical filter ;
		filterfac:units =  ;
	int64 twrite() ;
		twrite:long_name = interval for cfl writeout ;
		twrite:units = number of timesteps ;
	float64 tmax() ;
		tmax:long_name = total time of integration ;
		tmax:units = seconds ;
	float64 tavestart() ;
		tavestart:long_name = start time for averaging ;
		tavestart:units = seconds ;
	float64 taveint() ;
		taveint:long_name = time interval for accumulation of diagnostic averages ;
		taveint:units = seconds ;
	int64 ntd() ;
		ntd:long_name = number of threads used ;
		ntd:units = none ;

// global attributes:
	:title = Python Quasigeostrophic Model ;
	:institution =  ;
	:source = pyqg version: 0.3.0 ;
	:history =  ;
	:references =  ;
	:comment =  ;
}
```
![Screen Shot 2021-03-11 at 12 28 59 PM](https://user-images.githubusercontent.com/22622585/110828754-5daf2800-8265-11eb-82f7-e14a28268852.png)

**Comment by rabernat** [comment]
> it might be helpful to discuss how this will eventually integrate in with `pyqg`

First you will have to define a _function_ with the signature:
```python
def model_to_dataset(m):
```
which takes a pyqg model and returns an xarray dataset. You will put this function new module within pyqg called `xarray_output.py`. 

Then you will add a new method to the base pyqg `Model` class, near the end of [this file](https://github.com/pyqg/pyqg/blob/master/pyqg/model.py) that looks like the:
```python
    def to_dataset(self):
        from .xarray_output import model_to_dataset
        return model_to_dataset(self)
```

**Comment by rabernat** [comment]
Closed by #230.
